drivers/hwinfo/hwinfo_mcux_kinetis.c

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2018 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include <hwinfo.h>
+#include <string.h>
+#include <fsl_sim.h>
+
+#if defined(SIM_UIDH)
+#define HWINFO_DEVICE_ID_LENGTH_H 1
+#else
+#define HWINFO_DEVICE_ID_LENGTH_H 0
+#endif
+#if (defined(FSL_FEATURE_SIM_HAS_UIDM) && FSL_FEATURE_SIM_HAS_UIDM)
+#define HWINFO_DEVICE_ID_LENGTH_M 1
+#else
+#define HWINFO_DEVICE_ID_LENGTH_M 2
+#endif
+#define HWINFO_DEVICE_ID_LENGTH_L 1
+
+#define HWINFO_DEVICE_ID_LENGTH_TOTAL (HWINFO_DEVICE_ID_LENGTH_L + \
+				       HWINFO_DEVICE_ID_LENGTH_M + \
+				       HWINFO_DEVICE_ID_LENGTH_H)
+
+struct kinetis_uid {
+	u32_t id[HWINFO_DEVICE_ID_LENGTH_TOTAL];
+};
+
+ssize_t _impl_hwinfo_get_device_id(u8_t *buffer, size_t length)
+{
+	struct kinetis_uid dev_id;
+
+	dev_id.id[0] = SIM->UIDL;
+#if (defined(FSL_FEATURE_SIM_HAS_UIDM) && FSL_FEATURE_SIM_HAS_UIDM)
+	dev_id.id[1] = SIM->UIDM;
+#else
+	dev_id.id[1] = SIM->UIDML;
+	dev_id.id[2] = SIM->UIDMH;
+#if defined(SIM_UIDH)
+	dev_id.id[3] = SIM->UIDH;
+#endif /* SIM_UIDH */
```

Comment:
Rather than duplicating all the ifdef logic, why not just call `SIM_GetUniqueId`?

Reply:
It's more complicated because the struct is also modified by SIM_UIDH ...

Reply:
You're already copying the id twice. Why can't you do this:
```c
ssize_t _impl_hwinfo_get_device_id(u8_t *buffer, size_t length)
{
        sim_uid_t sim_uid;
        SIM_GetUniqueId(&sim_uid);                                                         

        if (length > sizeof(sim_uid)) {
                length = sizeof(sim_uid);
        }

        memcpy(buffer, &sim_uid, length);

        return length;
}
```

Reply:
Because of the reordering of high and low words.

Reply:
I'm not sure that matters, since it's still unique if you rearrange the words. Will you please add a note in the commit message explaining why you did not use SIM_GetUniqueId?

Reply:
Added the note